### DISCUSSION TITLE
How do I construct / define assets dynamically?

### DISCUSSION THREAD

**Discussion by sryza** [dagster-io/dagster#15119] [Q&A]
I learned this as the main way to define an asset in Dagster

```python
@asset
def asset1():
    ...
```

However, it doesn't make sense for me to have a separate definition like this for every one of my assets. There are too many and I can't edit Python every time I want to add one.  I need to be able to construct them more dynamically.

**Comment by sryza** [comment]
"Dynamically" can mean a few different things:

### I want to add assets without writing a separate Python function for each asset
In this scenario, the set of assets is mainly determined by the contents of your git repository, but you want to be able to add assets by just adding a config file or a line of config.

The most common way to implement this kind of pattern is with an "asset factory", which is just a plain Python function that returns an asset definition. You can invoke it a bunch of times with different parameters to construct a bunch of assets.

Here's an example of an asset factory that constructs a set of assets from a set of yaml-structured specs: https://github.com/dagster-io/dagster/discussions/11045.

### I want to add assets without reloading my definitions or re-deploying a container
In this scenario, the trigger for adding a new asset is something other than editing a file in a git repository. Instead, for example, it's a new data file arriving inside a directory.

In this case, we recommend using a partition to represent each file, instead of a separate asset for each file.

This blog post explains how this works: https://dagster.io/blog/dynamic-partitioning.

And here are the docs for dynamic partitions: https://docs.dagster.io/concepts/partitions-schedules-sensors/partitions#dynamically-partitioned-assets.

**Reply by nitindatta** [reply]
thanks @sryza that helps, This helps in dynamically creating assets during load time. However, assets are still loaded during startup when Dagster starts and loads files. Is there an API available for adding assets at runtime without restarting Dagster? I noticed a "Load" button in the Dagit UI. Does Dagster expose an API that I can call externally to add new assets during runtime?

In my use case, I am developing a no-code ETL product where users define data sources, targets, and transformation rules. I require the ability to create user-defined data sources as assets in Dagster, as I want to leverage Dagster's capabilities for lineage, dependencies, and scheduling jobs.

**Reply by sryza** [reply]
> Does Dagster expose an API that I can call externally to add new assets during runtime?

Dagster doesn't currently expose something like this. With the exception of dynamic partitions, as described above, the set of assets is fixed at load time.

**Reply by nitindatta** [reply]
thanks @sryza

**Reply by karakanb** [reply]
@sryza is it planned to have any such feature or is it not an interest at all? this could help in our scenario which is similar to describe above, we want to be able to define & update assets based on some external configuration that are being updated by multiple teams.

**Reply by sryza** [reply]
We're currently not planning to add graphql APIs that allow mutating the set of asset definitions, without reloading code locations. The assumption that asset definitions are loaded from code locations is fairly core to Dagster's architecture, so breaking that assumption would be a big change.

**Reply by karakanb** [reply]
The way I was imagining was more like a code location that reloads itself every X seconds rather than mutating the asset definitions via any API, so that I don't have to call APIs or restart code locations myself. 

Would sth like that fit into the way you envision the future of code locations?

**Reply by nitindatta** [reply]
hi @karakanb is there API available to reload asset definitions instead of restarting whole process/container.

**Reply by karakanb** [reply]
I don't know really, I haven't looked into it, although since the UI does it via a button I presume there must be an API somewhere, I'd start by tracking the network traffic from the UI. However, until it is officially provided as an API treat it more like an experiment, it might break / change at any version.

**Reply by jordandakota** [reply]
> We're currently not planning to add graphql APIs that allow mutating the set of asset definitions, without reloading code locations. The assumption that asset definitions are loaded from code locations is fairly core to Dagster's architecture, so breaking that assumption would be a big change.

Can a partition not represent an asset? Say I use dynamic partitions to load data, could I not set it up so each parquet landing in a folder triggers a dynamic "partition" but actually reloads a table in say duckdb?

**Reply by kinghuang** [reply]
@karakanb, did you ever implement dynamic assets by reloading code locations on a regular basis? I'm going down the same path. I see there's `ReloadRepositoryLocationMutation` and `ReloadWorkspaceMutation` GraphQL mutations. Are you doing that to reload code locations?
